autoload/lsp/internal/completion/documentation.vim

```diff
@@ -65,43 +68,113 @@ endfunction
+    " Ignore normalized documentation is empty.
+    let l:documentation = s:MarkupContent.normalize(l:completion_item['documentation'])
     if empty(l:documentation)
+        call s:close_floating_window(v:true)
         return
     endif
 
-    if exists('s:floating_win')
-        call s:floating_win.close()
+    " Add detail field if provided.
+    if type(get(l:completion_item, 'detail', v:null)) == type('')
+        if !empty(l:completion_item.detail)
+            let l:detail = s:MarkupContent.normalize({
+            \   'language': &filetype,
+            \   'value': l:completion_item['detail'],
+            \ })
+            let l:documentation = l:detail . "\n\n" . l:documentation
+        endif
     endif
 
-    " TODO: support markdown
-    let l:lines = lsp#utils#_split_by_eol(l:documentation)
-
-    let l:row = float2nr(a:event['row'])
-    let l:col = float2nr(a:event['col'])
-    let l:curpos = screenrow()
+    let l:doc_win = s:get_doc_win()
+    if empty(l:doc_win)
+        call s:close_floating_window(v:true)
+        return
+    endif
 
-    " TODO: reuse floating window/buffer??
-    let s:floating_win = s:FloatingWindow.new()
-    let l:bufnr = s:Buffer.create()
-    call setbufline(l:bufnr, 1, l:lines)
-    call s:floating_win.set_bufnr(l:bufnr)
+    " Update contents.
+    call deletebufline(l:doc_win.get_bufnr(), '^', '$')
+    call setbufline(l:doc_win.get_bufnr(), 1, lsp#utils#_split_by_eol(l:documentation))
+
+    " Calculate layout.
+    let l:size = l:doc_win.get_size({
+    \   'maxwidth': float2nr(&columns * 0.4),
+    \   'maxheight': float2nr(&lines * 0.4),
+    \ })
+    let l:pos = s:compute_position(a:event, l:size)
+    if empty(l:pos)
+        call s:close_floating_window(v:true)
+        return
+    endif
 
-    call s:floating_win.open({
-        \ 'row': 1,
-        \ 'col': 1,
-        \ 'width': 10,
-        \ 'height': 10,
+    if mode()[0] ==# 'i'
+        call l:doc_win.open({
+        \   'row': l:pos[0] + 1,
+        \   'col': l:pos[1] + 1,
+        \   'width': l:size.width,
+        \   'height': l:size.height,
         \ })
+    endif
+    call s:Window.do(l:doc_win.get_winid(), { -> s:Markdown.apply() })
+endfunction
+
+function! s:close_floating_window(force) abort
+    " Ignore `CompleteDone` if it occurred by `complete()` function because the popup menu will re-appear immediately.
+    let l:ctx = {}
+    function! l:ctx.callback(force) abort
+        let l:doc_win = s:get_doc_win()
+        if empty(l:doc_win)
+            return
+        endif
+
+        if !pumvisible() || mode()[0] !=# 'i' || a:force
+            call l:doc_win.close()
+        endif
+    endfunction
+    call timer_start(1, { -> l:ctx.callback(a:force) })
 endfunction
 
-function! s:close_floating_window() abort
-    if exists('s:floating_win')
-        call s:floating_win.close()
+function! s:compute_position(event, size) abort
+    let l:col_if_right = a:event.col + a:event.width + 1 + (a:event.scrollbar ? 1 : 0)
```

Comment:
not required now and can do later. but we should port the logic from neovim doc floating. it solves lot of overlapping issue and also includes docking. we can slowly do this as separate PR.

Reply:
One way to try out is by trying to open popups around corner or in small splits and windows or bottom and top and see if it overlaps or not. there is also dock support by using popup which some folks that have tall monitors use but we can fix this later.

Reply:
_I personally prefer to avoid displaying popup-menu if `left/right` has no enough space for both._

But agree with you. we can improve it later.